File: day177/max_or.py
# https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets/description/

class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        max = reduce(operator.or_, nums, 0)
        res = 0
        for subset in chain.from_iterable(combinations(nums, r) for r in range(len(nums)+1)):
            res += 1 if reduce(operator.or_, subset, 0) == max else 0 
        return res
        

File: day177/max_or.rs
// https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets/description/

pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
    let max_or = nums.iter().fold(0, |acc, num| acc | num);
    let total_subsets = 1 << nums.len();
    (0..total_subsets).fold(0, |res, mask| {
        let cur_or = nums.iter().enumerate().fold(0, |acc, (index, num)| {
            if ((mask >> index) & 1) == 1 {
                acc | num
            } else {
                acc
            }
        });

        if cur_or == max_or {
            res + 1
        } else {
            res
        }
    })
}
